[OpenMP][Offload] Handle `present/to/from` when a different entry did `alloc/delete`.

abhinavgaba opened the pull request:
OpenMP allows cases like the following:

```c
  int *p1, *p2, x;
  p1 = p2 = &x;
  ...
  #pragma omp target_exit_data map(delete: p1[:]) from(p2[0])
```

Which means, when the runtime encounters the `from` entry, the ref-count may not be zero, but it will go down to zero at the end of the current construct, which should cause the "from" transfer to happen.

Similarly, a user may have:

```c
  struct S {
    int *p;
  };

  #pragma omp declare_mapper (id1: S s) map(s.p) map(present, alloc: s.p[0:10])
  #pragma omp declare_mapper (id2: S s) map(s.p, s.p[0:10])

  S s1;

 // present-check should fail here
 #pragma omp target_enter_data map(alloc: s.p[0:10]) map(mapper(id1), to: s)
 // "to" should be honored here
 #pragma omp target_enter_data map(alloc: s.p[0:10]) map(mapper(id2), to: s)
```

Where the allocation happens before the "to" entry is encountered by the runtime. Or, an allocation happens before a "present" entry is encountered.

To handle cases like this, we need to use the state information of previously seen new allocations, deletions, "from" entries, when honoring `to`/`from`/`present` map entries.

abhinavgaba committed:
[OpenMP][Offload] Handle for non-memberof present/to/from entries irrespective of order.

For cases like:
```c
  map(alloc: x) map(to: x)
```
If the entry of `map(to: x)` is encountered after the entry for
`map(alloc:x)`, we still want to do a data-transfer even though the
ref-count of `x` was already 0, because the new allocation for `x`
happened as part of the current directive.

Similarly, for:
```c
  ... map(alloc: x) map(from: x)
```

If the entry for `map(from:x)` is encountered before the entry for
`map(alloc:x)`, we want to do a data-transfer even though the
ref-count was not 0 when looking at the `from` entry, because by the end of
the directive, the ref-count of `x` will go down to zero.

And for:
```c
  ... map(from : x) map(alloc, present: x)
```

If the "present" entry is encountered after the "from" entry, then it becomes
a no-op, as the "from" entry will do an allocation if no match was found.

In this PR, these are handled by the runtime via the following:
* For `to` and `present`, we also look-up in the existing table where we tracked
new allocations when making the decision for the entry.
* For `from`, we keep track of any deferred data transfers and when the
ref-count of a pointer goes to zero, see if there were any previously
deferred `from` transfers for that pointer.

This can be done in the compiler, and that would avoid any runtime
overhead, but it would require creating two separate offload struct entries
for the entry and exit mappings (even for the `target` construct),
with properly decayed maps, and either:
(1) sorted in order of:
* `present > to > ...` for the implied `target enter data`; and
* `from > ...` for the `target exit data`
e.g.

```c
  #pragma omp target map(to: x) map(present, alloc: x) map(always, from: x)
  // has to be broken into:
  // from becomes alloc on entry:
  //  #pragma omp target enter data map(present, alloc: x)
  //                                map(to: x)
  //                                map(alloc: x)
  //
  //  "present" and "to" just "decay" into "alloc"
  //  #pragma omp target exit data map(always, from: x)
  //                               map(alloc: x)
  //                               map(alloc: x)
```
Or,
(2) Merged into one entry each on the `target enter/exit data`
directives.

```c
  #pragma omp target map(to: x) map(present, alloc: x) map(always, from: x)
  // has to be broken into:
  // from becomes alloc on entry:
  //  #pragma omp target enter data map(present, to: x)
  //
  //  "present" and "to" just "decay" into "alloc"
  //  #pragma omp target exit data map(always, from: x)
```

The number of entries on the two would need to stay the same on the two to avoid
ref-count mismatch.

(1) would be simpler, but won't likely work for cases like:
```c
  ... map(delete: x) map(from:x)
```
as there is no clear "winner" between the two. So, for such cases, the compiler
would likely have to do (2), which is the cleanest solution, but will take
longer to implement. For EXPR comparisons, it can build-upon the
`AttachPtrExprComparator` that was implemented as part of #153683,
but that should probably wait for the PR to be merged to avoid
conflicts.

Another alternative is to sort the entries in the runtime, which may be
slower than on-demand lookups/updates that this PR does, because we
always would be doing this sorting even when not needed, but may be
faster in others where the constant-time overhead of map/set
insertions/lookups becomes too large because of the number of maps. But
that will still have to worry about the `from` + `delete` case.

abhinavgaba committed:
Fix 'from' + 'delete', and multiple 'always,from' entries.

dreachem commented:
@abhinavgaba 

> Another thing to think of, but maybe from the spec's perspective, is that should the modifiers from a `from/to` clause be completely ignored on `enter/exit_data`?
> 
> For the following:
> 
> ```c
> int x;
> 
> int main(void) {
>   #pragma omp target map(delete, to: x) map(present, from: x)
>   ;
> }
> ```
> 
> @dreachem, @mjklemm, is this expected to run OK? If we decay `from` into `storage`, and then apply it to `enter_data`, then it would seem to imply that it's equivalent to doing `map(present, to: x)`, in which case, this will cause a runtime-error-termination.
> 
> ```c
> // Applying "present" to "enter_data" would cause runtime-error
> // (5.1)
> #pragma omp target_enter_data(present, to: x)
> ...
> // (5.2)
> #pragma omp target_exit_data(present, delete, from: x)
> ```
> 
> But the user may be expecting to only do a present-check on `exit_data` here, which would means we need to ignore any modifiers from a `from/to` clause that's decayed into `storage` on `target_enter/exit_data`.
> 
> ```c
> // Ignoring modifiers from the decayed-to-storage clause.
> // (6.1)
> #pragma omp target_enter_data(to: x) // delete ignored for enter_data
> ...
> // (6.2)
> #pragma omp target_exit_data(present, from: x)
> ```
> 
> But if we do so, then `delete` gets ignored on `target_exit_data`, since it came from a decayed `to`. Maybe we only special-case present and not apply it to decayed-to-storage maps.

So, you are describing `target` as if it there is a `target_enter_data` on entry and a `target_exit_data` on exit, but this isn't how the specification describes it. That's the description for `target_data`. 

For `target`, a `map(delete, to: x)` clause has the effect of `map(to: x)` on entry and `map(delete,storage:x)` on exit, and a `map(present,from:x)` clause has the effect of a `map(present,storage:x)` on entry and a `map(from:x)` on exit.  So, putting it together, for that `target` construct you effectively get the equivalent of `map(present, to: x)` on entry and `map(delete,from:x)` on exit. 

Let's suppose the construct was instead `target_data map(delete,to: x) map(present,from: x)`. In that case, we describe it as equivalent to:

```
// note that "delete" on target_enter_data is just ignored
#pragma omp target_enter_data map(delete,to: x) map(present,storage: x)
#pragma omp task if(0) // a merged "sharing task"
;
#pragma omp target_exit_data map(delete,storage: x) map(present,from: x)
```

Here, it looks like we hit a possible unintended discrepancy. When we decided to make `target_data` composite, I'm not sure we accounted for the behavior of the `present` modifier possibly changing. Since there are now 2 data-mapping regions each with their own entry and exit, as opposed to a single `target_data` region with one entry and one exit, there are two points at which the `present` modifier can have an effect.  I don't think this behavior is unreasonable for `target_data`, but just wanted to call it out.

abhinavgaba committed:
Handle cases like map(delete:p[:]) map(from:q[0:1]).

abhinavgaba committed:
Clang-format

abhinavgaba committed:
Add two mapper tests, avoid a rendant map lookup.

abhinavgaba committed:
Change Deferred to Skipped, since that's more accurate

abhinavgaba committed:
Clang-format

abhinavgaba committed:
Fix some tests.

abhinavgaba commented on offload/test/mapping/map_ordering_ptee_tgt_alloc_mapper_alloc_from_to.c:
This is an interesting case.

@dreachem Is it legal to have `map(alloc: s1.p[0:10])` on a `target` directive, and then have a `map(to: s1.p[0:10])` contributed via the mapper?

I'm thinking it's legal since they are the "same list-item".

Follow-up: If it is legal, would it be illegal if we had:

```c
pragma omp target map(alloc : x) map(mapper(my_maper): s1)
... 
```

abhinavgaba committed:
Support case of FROM pointer falling in the middle of the last entry.

abhinavgaba committed:
Support TO/PRESENT where the begin address falls in the middle of a prior new allocation.

abhinavgaba committed:
Support multiple FROMs falling within the same LAST entry.

abhinavgaba reviewed (commented):
I updated the checks to handle cases where the from entries don't start with the same address as the last entry that causes deletion, and the case when multiple skipped from entries fall within the same entry being deleted, like https://github.com/llvm/llvm-project/pull/165494/files#diff-eb82dd071c1acec10b994ecb40b122e2b586f147e8f682a32361a181a28ab77c.

I think OpenMP 6.0 may have went too far in allowing overlaps with any assumed-size/implicit use_device_addr maps. I'll create a spec ticket to discuss it further.

abhinavgaba committed:
Merge remote-tracking branch 'upstream/main' into fix-map-to-from-ordering

abhinavgaba committed:
Address Robert's review comments.

abhinavgaba committed:
Clean-up some comments, tests.

abhinavgaba committed:
Remove two tests that were useful, but not strictly OpenMP compliant.

abhinavgaba committed:
Merge remote-tracking branch 'upstream/main' into fix-map-to-from-ordering

abhinavgaba committed:
Clang-format fixes.

abhinavgaba committed:
Make from+from overlap case more robust.

Also improve robustness of multiple FROM entries of different sizes
but same starting address.

ro-i commented:
works on my machine, thanks!

ro-i commented:
(give me a day or so to look at the new changes you made)

ro-i commented on offload/test/mapping/map_ordering_ptee_tgt_data_alloc_tgt_mapper_present_delete_from_to.c:
```diff
@@ -0,0 +1,48 @@
+    printf("%d\n", s1.p[1]); // CHECK: 222
+  }
+  // clang-format off
+  // DEBUG: omptarget --> Found skipped FROM entry: HstPtr=0x[[#%x,HOST_ADDR:]] size=[[#%u,SIZE:]] within region being deleted
```
```suggestion
  // DEBUG: omptarget --> Found skipped FROM entry: HstPtr=0x[[#%x,HOST_ADDR:]] size=[[#%u,SIZE:]] within region being released
```
since you updated the wording

abhinavgaba commented on offload/include/OpenMP/Mapping.h:
```diff
@@ -495,28 +495,118 @@ struct AttachMapInfo {
+  /// Returns the matching entry if found, otherwise std::nullopt.
+  std::optional<std::pair<void *, int64_t>>
+  wasTransferredFrom(void *Ptr, int64_t Size) const {
+    uintptr_t CheckBegin = reinterpret_cast<uintptr_t>(Ptr);
```
We get errors like this with static_cast from void* to intptr/uintptr:

```
Mapping.h:562:28: error: static_cast from 'void *' to 'uintptr_t' (aka 'unsigned long') is not allowed
```

abhinavgaba committed:
Fix a CHECK in one test.

abhinavgaba reviewed (commented):
Thanks for the review!

abhinavgaba committed:
Add missing fopenmp-version to test.

abhinavgaba committed:
Add XFAIL intelgpu to new tests.

ro-i reviewed (approved):
LGTM afaict
